### ISSUE TITLE
Make executing nohlsearch from user functions possible.

### ISSUE THREAD

**Issue by purpleP** [neovim/neovim#5581]
It is really surprising that such a trivial thing as automatically disable highlight after some time when say cursor moved is impossible with vim and need to use [plugins](https://github.com/romainl/vim-cool/) for that, that remaps half of the keyboard and wouldn't work if I say remapped some search related keys. If `noh` could be called from user functions this would be as trivial as

`au CursorMoved call timer_start(2000, 'DisableHighlight')`

**Comment by chrisbra** [comment]
Am 2016-11-10 14:30, schrieb purpleP:
> It is really surprising that such a trivial thing as automatically
> disable highlight after some time when say cursor moved is impossible
> with vim and need to use plugins [1] for that, that remaps half of the
> keyboard and wouldn't work if I say remapped some search related keys.
> If noh could be called from user functions this would be as trivial as
> 
> 
> au CursorMoved call timer_start(2000, 'DisableHighlight')

its already possible with feedkeys.

**Comment by purpleP** [comment]
@chrisbra 
I still haven't figured out completely what feedkeys is doing and how it works, but your [suggestion](https://gist.github.com/chrisbra/737b4b0487a7798af6f179bda9ce8c09#file-searchhighlight-vim) isn't working properly in operator pending mode as all workarounds that I know are.

**Comment by chrisbra** [comment]
See the help `:h feedkeys()`

**Comment by purpleP** [comment]
@chrisbra Yeah, I've read it. I still not sure how to avoid accidental strings that passed to feedkeys inserted into buffer. Do you know how to do that?

**Comment by purpleP** [comment]
I think this is already have been discussed, but what I really need is a way to react to pressed keys without remappings.
Specifically what I want is to disable highlight if I don't need to see search matches. Which translates to following behaviour.
1. If I searched with `/` or `?` as soon as cursor moved (except when it moved because of `n` or `N` highlight should turn off.
2. If I searched with `*` or `#` as soon as cursor moved (except the first move to next occurrence of the word or `n` or `N`) highlight should turn off.
3. If I use search in operator pending mode don't highlight matches at all in the first place.

What bugs me is this is full specification of what I want, not just some heuristic that would cover most of the cases, yet is seems to be impossible to express this simple idea in vim.

**Comment by justinmk** [comment]
> what I really need is a way to react to pressed keys without remappings.

I have a work-in-progress local branch that enables something like that.

In the meantime regarding `:hlsearch` specifically, the limitation described at `:hlsearch` _could_ be worked around by a medium-risk refactoring. Patches welcome.

**Comment by jamessan** [comment]
Isn't this what `v:hlsearch` is for?  `let v:hlsearch = 0` if you want to disable the highlight, and `let v:hlsearch = 1` if you want to enable it (if the user has `'hlsearch'` set).

**Comment by purpleP** [comment]
@jamessan It actually works but what I've summarised what I really want from vim above. The way I see it to make vim do what I want I would have to either reimplement vim search functionality using vimscript (and I've just tried to play with `search` function and I can't find how to make it highlight matches) or if what I've asked initially would be possible this could be as simple as this

```python
# python pseudo-code
def onKeyPressed(mode, key):
     if mode is OPERATOR_PENDING_MODE and key is in {'/', '?'}:
        highlight_search_off()
     elif highlight_on() and key is not in {'n', 'N'}:
        nohl()
```

**Comment by purpleP** [comment]
You know what's funny, guys? Emacs with evil mode seems to be doing exactly what I want in this case. It fails in others like for some reason <C-w><C-h> doesn't work, only <C-w>h etc for switching windows.

**Comment by justinmk** [comment]
Yes, there's no fundamental reason this couldn't be fixed, and I wouldn't expect evil-mode to replicate this bug. This bug exists because of a Vim implementation detail.

**Comment by purpleP** [comment]
@jamessan @justinmk 
Hey, it started to bug me again and I've almost done everything I've wanted.

```vim
fu! HlSearch()
    let s:pos = match(getline('.'), @/, col('.') - 1) + 1
    echom s:pos . '|' . col('.')
    if s:pos != col('.')
        call StopHL()
    endif
endfu

fu! StopHL()
    if !v:hlsearch || mode() isnot 'n'
        return
    else
        let cmd="\<c-\>\<c-n>:let v:hlsearch=0\<cr>\<c-l>"
        sil call feedkeys(cmd, 'n')
    endif
endfu

augroup SearchHighlight
au!
    au CursorMoved * call HlSearch()
    au InsertLeave * call StopHL()
augroup end
```

I have only one problem now - I can't use `InsertEnter` event with`StopHL` in operator pending mode because it messes my buffer (which is probably expected since it emulates me typing on keyboard and I don't want that emulation in insert mode). Which got me thinking - what if there would be an `InsertEnterPre` event or `NormalLeavepre`. As you've probably already guessed the idea is to use `feedkeys` just before going to insert mode but still in normal mode.

**Comment by jamessan** [comment]
Does it help if you use `feedkeys(cmd, 'in')`?

**Comment by ZyX-I** [comment]
@purpleP Try using `<expr>` mappings instead:

```VimL
noremap <expr> <Plug>(StopHL) execute('nohlsearch')[-1]
noremap! <expr> <Plug>(StopHL) execute('nohlsearch')[-1]

function s:StopHL()
    if !v:hlsearch
        return
    else
        call feedkeys("\<Plug>(StopHL)", "m")
    endif
endfunction
```

. The main idea is that you need to use `<expr>` mapping with expression with a side-effect (running `nohlsearch`) which evaluates to an empty string (thus not adding anything to the buffer or disturbing your input; except for the cases when you partially typed a mapping before `s:StopHL()` interferred). This should work with any mode.

**Comment by purpleP** [comment]
@ZyX-I Wow, thanks a lot! Thanks to you I've just removed the main frustration from vim. Now I don't need to turnoff highlight after searching something and it won't bother me in operator pending mode. Previously I've just disabled search highlight at all, because it turned out that the cons (needing to turn it off every time) are just outweigh the pros (sometimes it's not so obvious where is the next match). But with recent neovim `inccommand` feature it's super useful in search and replace. And I've started to turn it on before every `%s` command.

**Comment by bounceme** [comment]
```
$ nvim i.c "+au insertenter * echom v:hlsearch"
$ nvim -i NONE i.c "+au insertenter * echom v:hlsearch"
```
when using @romainl ' s `cool` plugin.

is it a bug? when starting nvim with `-i NONE` the result is 1 and `nohlsearch` is echoed

**Comment by bounceme** [comment]
after this commit: https://github.com/romainl/vim-cool/commit/df11d4cd43980f6f4a94cd24a272ac3397e1dfff

**Comment by adamency** [comment]
@ZyX-I Thanks a lot for your suggestion. However would you mind explaining the point and effect of using `[-1]` after the `execute` ? Cause when testing with or without in a simple mapping that I call manually, I am failing to see any difference
